### PR TITLE
Introduce incrementalTool option to support gradle

### PR DESCRIPTION
**Pull request by semihokur**

### CODE REVIEW

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs [hunk 1]
```diff
@@ -310,6 +328,12 @@ public sealed class SandboxedProcessPipExecutor : ISandboxedProcessFileStorage
             }
 
             m_vmInitializer = vmInitializer;
+
+            m_incrementalToolFragments = new List<string>(
```
Comment (almiliMSFT) [inline Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: So cryptic with that `??` operator hidden in the back.  :)

This would look less cryptic to me:
```csharp
m_incrementalToolFragments = (incrementalTools ?? Enumerable.Empty<string>())
  .Select(...)
  .ToList();
```

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs [hunk 2]
```diff
@@ -2052,7 +2076,7 @@ private void AddStaticFileDependencyToFileAccessManifest(FileArtifact dependency
                 // to it. Explicitly block this, since we want inputs to not be written. Observe we already know 
                 // this is not a rewrite.
                 mask: m_excludeReportAccessMask &
-                      ~FileAccessPolicy.AllowRealInputTimestamps &
+                      (NoFakeTimestamp ? ~FileAccessPolicy.Deny : ~FileAccessPolicy.AllowRealInputTimestamps) &
```
Comment (almiliMSFT) [inline Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: The expression
```csharp
NoFakeTimestamp ? ~FileAccessPolicy.Deny : ~FileAccessPolicy.AllowRealInputTimestamps
```
seems to occur at least 4 times, so maybe extract it into a property.

File: Public/Src/FrontEnd/Script/Ambients/Transformers/AmbientTransformer.Process.cs [hunk 3]
```diff
@@ -1179,6 +1181,17 @@ private void ProcessUnsafeOptions(Context context, ProcessBuilder processBuilder
...
                 processBuilder.Options |= Process.Options.AllowPreserveOutputs;
+
+                if (context.FrontEndHost.Configuration.Sandbox.PreserveOutputsForIncrementalTool)
+                {
+                    processBuilder.Options |= Process.Options.IncrementalTool;
+                }
+            }
+
+            // UnsafeExecuteArguments.incrementalTool
+            if (Converter.ExtractOptionalBoolean(unsafeOptionsObjLit, m_unsafeIncrementalTool) == true)
```
Comment (dannyvv) [inline Public/Src/FrontEnd/Script/Ambients/Transformers/AmbientTransformer.Process.cs]: Don't forget to add the option to the type in the dsc file at:

[https://github.com/microsoft/BuildXL/blob/master/Public/Sdk/Public/Transformers/Transformer.Execute.dsc#L216](https://github.com/microsoft/BuildXL/blob/master/Public/Sdk/Public/Transformers/Transformer.Execute.dsc#L216)

File: Public/Src/Engine/Processes/ReportedFileAccess.cs [hunk 4]
```diff
@@ -218,20 +218,7 @@ public string ShortDescribe(PathTable pathTable)
...
-                    case RequestedAccess.All:
-                    case RequestedAccess.Write:
-                    case RequestedAccess.ReadWrite:
-                        return true;
-                    default:
-                        return false;
-                }
-            }
-        }
+        public bool IsWriteViolation => (RequestedAccess & RequestedAccess.Write) != 0;
```
Comment (dannyvv) [inline Public/Src/Engine/Processes/ReportedFileAccess.cs]: Since we are moving to NetCore and away from Net472 we can start using HasFlag. Since it doesn't have boxing cost in dotnetcore..

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs [hunk 5]
```diff
@@ -4030,5 +4065,40 @@ private bool ShouldPreserveDeclaredOutput(AbsolutePath path, HashSet<AbsolutePat
...
+            }
+
+            bool result;
+            if (m_incrementalToolMatchCache.TryGetValue(toolPath, out result))
+            {
+                return result;
+            }
+
+            result = false;
+            foreach (var incrementalToolSuffix in m_incrementalToolFragments)
```
Comment (dannyvv) [inline Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: since the fragments are relative paths, should we put a path separator in front of them before we check a prefix to avoid:

fragmentname: "foo.exe"
toolPath: "c:\somefoo.exe"

this would be a hit where it shouldn't no?

File: Public/Src/Utilities/Configuration/IRootModuleConfiguration.cs [hunk 6]
```diff
@@ -23,5 +23,11 @@ public interface IRootModuleConfiguration : IModuleConfiguration
         /// </summary>
         [CanBeNull]
         IReadOnlyList<RelativePath> SearchPathEnumerationTools { get; }
+
+        /// <summary>
+        /// List of path fragments of tools using its own incrementality
+        /// </summary>
+        [CanBeNull]
+        IReadOnlyList<RelativePath> IncrementalTools { get; }
```
Comment (dannyvv) [inline Public/Src/Utilities/Configuration/IRootModuleConfiguration.cs]: Curious why this is on the root configuration. Should we do it per pip? Users need to set a bit on the pip anyway...

Comment (semihokur) [reply Public/Src/Utilities/Configuration/IRootModuleConfiguration.cs]: It is very similar to searchpathenumeration tools. It is global.

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs (hunk 5 above)
Comment (semihokur) [reply Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: Yes, I already put a path separator in the constructor, please see line #335 in the same class.

File: Public/Src/App/Bxl/Args.cs [hunk 7]
```diff
@@ -864,6 +864,10 @@ public bool TryParse(string[] args, PathTable pathTable, out ICommandLineConfigu
                         OptionHandlerFactory.CreateOption(
                             "tempDirectory",
                             opt => layoutConfiguration.TempDirectory = CommandLineUtilities.ParsePathOption(opt, pathTable)),
+                        OptionHandlerFactory.CreateBoolOption(
+                            "temporary_PreserveOutputsForIncrementalTool",
```
Comment (narasamdya) [inline Public/Src/App/Bxl/Args.cs]: >temporary_ [](start = 29, length = 10)

It should no longer be temporary.

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs [hunk 8]
```diff
@@ -1875,7 +1899,7 @@ private void TryPrepareFileAccessMonitoringForPip(Process pip, HashSet<AbsoluteP
                         if (!directoryOutputIds.Contains(directory.Path))
                         {
                             // Directories here represent inputs, we want to apply the timestamp faking logic
-                            var mask = ~FileAccessPolicy.AllowRealInputTimestamps;
+                            var mask = NoFakeTimestamp ? ~FileAccessPolicy.Deny: ~FileAccessPolicy.AllowRealInputTimestamps;
```
Comment (narasamdya) [inline Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: >: [](start = 79, length = 1)

nit: space

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs [hunk 9]
```diff
@@ -2977,8 +3005,15 @@ private bool CheckIfPathContainsSymlinks(AbsolutePath path)
                             // To treat the paths as file probes, all accesses to the path must be the probe access.
                             isProbe &= access.RequestedAccess == RequestedAccess.Probe;
 
+                            if (access.RequestedAccess == RequestedAccess.Probe
+                                && access.Operation == ReportedFileOperation.ReparsePointTarget
```
Comment (narasamdya) [inline Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: >access.Operation == ReportedFileOperation.ReparsePointTarget [](start = 35, length = 60)

Why do we constraint it for ReparsePointTarget?

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs [hunk 10]
```diff
@@ -4030,5 +4065,40 @@ private bool ShouldPreserveDeclaredOutput(AbsolutePath path, HashSet<AbsolutePat
 
             return false;
         }
+
+        private bool IsIncrementalToolAccess(ReportedFileAccess access)
+        {
+            if (!IsIncrementalPreserveOutputPip)
```
Comment (narasamdya) [inline Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: >!IsIncrementalPreserveOutputPip [](start = 16, length = 31)

Perhaps: `!IsIncrementalPreserveOutputPip || (m_incrementalToolFragments.Count == 0))`

File: Public/Src/Utilities/Configuration/EngineEnvironmentSettings.cs [hunk 11]
```diff
@@ -112,6 +112,10 @@ public static class EngineEnvironmentSettings
         // $Rename: Due to telemetry backend scripts this cannot be renamed to BuildXL
         public static readonly Setting<string> BuildXLUserName = CreateSetting("BUILDXL_USERNAME", value => value);
 
+        /// <summary>
+        /// Defines whether file sizes and hashes should be exported into the json graph
+        /// </summary>
+        public static readonly Setting<bool> RuntimeCacheMissAllPips = CreateSetting("BuildXLRuntimeCacheMissAllPips", value => value == "1");
```
Comment (narasamdya) [inline Public/Src/Utilities/Configuration/EngineEnvironmentSettings.cs]: Is this related? If not, please make a separate PR.

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs (hunk 1 above)
Comment (semihokur) [reply Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: Agreed; however, yours give some compilation errors. To fix them, it will get things complicated. i just added if/else branch.

---
In reply to: [310279032](https://github.com/microsoft/BuildXL/pull/697#discussion_r310279032) [](ancestors = 310279032)

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs (hunk 5 above)
Comment (semihokur) [reply Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: please see line#335

---
In reply to: [310289947](https://github.com/microsoft/BuildXL/pull/697#discussion_r310289947) [](ancestors = 310289947)

File: Public/Src/App/Bxl/Args.cs (hunk 7 above)
Comment (semihokur) [reply Public/Src/App/Bxl/Args.cs]: it is still temporary. we will get it removed when they pass processoptions.incrementaltool

---
In reply to: [310330852](https://github.com/microsoft/BuildXL/pull/697#discussion_r310330852) [](ancestors = 310330852)

File: Public/Src/Engine/Scheduler/Tracing/RuntimeCacheMissAnalyzer.cs [hunk 12]
```diff
@@ -227,7 +227,8 @@ private bool IsCacheMissEligible(PipId pipId)
                 return false;
             }
 
-            if (m_changedPips.WasVisited(pipId.ToNodeId()))
+            if (!EngineEnvironmentSettings.RuntimeCacheMissAllPips 
+                && m_changedPips.WasVisited(pipId.ToNodeId()))
```
Comment (narasamdya) [inline Public/Src/Engine/Scheduler/Tracing/RuntimeCacheMissAnalyzer.cs]: Not sure if this is related. If possible, please put in a separate PR.

**Comment by narasamdya** [comment]
Some unit tests may be helpful :-)

File: Public/Src/Engine/Scheduler/Artifacts/FileContentManager.cs [hunk 13]
```diff
@@ -2267,13 +2269,25 @@ private static PipOutputOrigin GetPipOutputOrigin(ContentMaterializationOrigin o
...
                             bool shouldDiscoverContentOnDisk =
                                 Configuration.Schedule.ReuseOutputsOnDisk ||
                                 !Configuration.Schedule.StoreOutputsToCache ||
                                 isPreservedOutputFile;
 
+                            var assumeContentUpToDate = isPreservedOutputFile &&
+                               Configuration.Distribution.BuildRole == DistributedBuildRoles.None
+                                && m_preservedUpToDateCacheLookupFiles.Contains(fileArtifact);
+
+                            if (materializingOutputs && state != null && assumeContentUpToDate)
```
Comment (narasamdya) [inline Public/Src/Engine/Scheduler/Artifacts/FileContentManager.cs]: >assumeContentUpToDate [](start = 73, length = 21)

Can you put some comment why we need to consider `assumeContentUpToDate` here?

Comment (semihokur) [reply Public/Src/Engine/Scheduler/Artifacts/FileContentManager.cs]: i decided to remove that optimization.
